### DISCUSSION TITLE
Group Totals

### DISCUSSION THREAD

**Discussion by mrajalko** [QuestPDF/QuestPDF#163] [Q&A]
Is there any example for creating group headers (repeating group headers on each page if group transcends multiple pages) and group footer to show the totals.

**Comment by MarcinZiabek** [comment]
Hello 😁 I am afraid that right now it is not possible. 

However, this week, I am planning to prepare an experimental release that should be very useful in this requirement. More details about the feature can be [found here](https://github.com/QuestPDF/QuestPDF/discussions/39). Please take a look and let me know what do you think. This is an advance use case and I am not sure how to shape the API. All feedback is always welcome 😁

**Reply by mrajalko** [reply]
Hello @MarcinZiabek Thanks for your response.

I have gone through the Page Total Requirement but it is different from my Requirement.
Most of business reports are made such that they are ordered in groups and after change of group the totals are printed and at the end of report grand total is printed.

this feature would make this framework more appealing to all developers like me who are doing this by HTMLTOPDF converter.

**Reply by MarcinZiabek** [reply]
Hmm, I am not sure what you requirement is exactly 🤣 Please take a look [at this element](https://www.questpdf.com/documentation/api-reference.html#decoration). If your content spans multiple pages, it adds header and footer for that content on EVERY page. If you are looking for something else, please describe your idea 😁

**Reply by mrajalko** [reply]
Yes the content spans multiple pages and I want to check if the value of a given column in a table changes and print totals and start a new group. 

I think implementing section can be useful in my case. if Only section can start and end in middle of the page.
if section start means a new page always I can use it.

It would be helpful be you can point to a good section example.

**Reply by MarcinZiabek** [reply]
Would you like to prepare an example with more detailed description? Possibly with sketches/mockups? Your requirement is still unclear to me and it is hard to help I don't know what "group/total/change" means exactly in your context. 😥

**Reply by mrajalko** [reply]
[CustomerLedger.pdf](https://github.com/QuestPDF/QuestPDF/files/8449956/CustomerLedger.pdf)

This is the output that I have created using Quest PDF.
I would share the code files shortly.
I am still struggling to repeat name of the group on next page if the content spans more than one page.

**Reply by mrajalko** [reply]
using InternalApp.Internal.Entities;
using QuestPDF.Drawing;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;
using System.Linq;

namespace InternalApp.Internal
{
    public class QuestCustomerDocument : IDocument
    {
        public QuestCustomerModel Model { get; }

        public static int FONT_SIZE_GRAND_TOTAL = 12;
        public static int FONT_SIZE_SUB_TOTAL = 12;
        public static int FONT_SIZE_FIELD_TITLE = 12;
        public static int FONT_SIZE_FIELD_VALUE = 12;
        public QuestCustomerDocument(QuestCustomerModel model)
        {
            Model = model;
        }
        public interface IComponent
        {
            void ComposeImage(IContainer container);
        }
        public DocumentMetadata GetMetadata() => DocumentMetadata.Default;

        public void Compose(IDocumentContainer container)
        {
            container.Page(page =>
            {
                page.Margin(50);

                page.Background() // <- new slots, Foreground and Background
                    .AlignMiddle()
                    .AlignCenter()
                    .Text("Digisure")
                    .FontSize(64)
                    .FontColor(Colors.Blue.Lighten5);

                page.Header().Element(ComposeImage);
                page.Header().Element(ComposeHeader);
                page.Content().Element(ComposeContent);

                page.Footer().AlignCenter().Text(x =>
                {
                    x.CurrentPageNumber();
                    x.Span(" / ");
                    x.TotalPages();
                });
                page.Size(PageSizes.A4.Landscape());
            });
        }
        void ComposeImage(IContainer container)
        {
            container.Image(Model.LogoData);
        }
        void ComposeHeader(IContainer container)
        {
            var titleStyle = TextStyle.Default.FontSize(20).SemiBold().FontColor(Colors.Blue.Medium);

            container.Row(row =>
            {
                row.RelativeItem().Column(column =>
                {
                    column.Item().Text($"{Model.Title}").Style(titleStyle);

                    //column.Item().Text(text =>
                    //{
                    //    text.Span("REPORT DATE: ").SemiBold();
                    //    text.Span($"{Model.ReportDate:d}");
                    //});
                });
                //row.ConstantItem(100).Height(50).Element(ComposeImage);
            });
        }
        void ComposeContent(IContainer container)
        {
            container.PaddingVertical(40).Column(column =>
            {
                column.Spacing(5);

                column.Item().Element(ComposeTable);
                column.Item().Element(ComposeGrandTotal);

                if (!string.IsNullOrWhiteSpace(Model.Title)) column.Item().PaddingTop(25);//.Element(ComposeComments);
            });
        }
        void ComposeTable(IContainer container)
        {
            container.Table(table =>
            {
                // step 1
                table.ColumnsDefinition(columns =>
                {
                    columns.RelativeColumn(1);
                    columns.RelativeColumn(3);
                    columns.RelativeColumn(2);
                    columns.RelativeColumn(3);

                    columns.RelativeColumn(2);
                    columns.RelativeColumn(3);
                    columns.RelativeColumn(2);
                    columns.RelativeColumn(2);

                    columns.RelativeColumn(2);
                    columns.RelativeColumn(2);
                    columns.RelativeColumn(2);
                });

                table.Header(header =>
                {
                    //=========== DISPLAYING FIELD TITLE ===================
                    string[] titles = {   "Sr."
                                        , "Invoice No."
                                        , "Invoice Date"
                                        , "Order No."

                                        , "Order Date"
                                        , "Transaction No."
                                        , "Payment Date"
                                        , "Contract Value"

                                        , "Pay Amount"
                                        , "TDS Amount"
                                        , "Balance"};

                    for (int i = 0; i < 11; i++)
                    {
                        if (i > 6)
                        {
                            header.Cell().AlignRight().Text(titles[i]).FontSize(FONT_SIZE_FIELD_TITLE).SemiBold().FontColor(Colors.Black);
                        }
                        else
                        {
                            header.Cell().AlignLeft().Text(titles[i]).FontSize(FONT_SIZE_FIELD_TITLE).SemiBold().FontColor(Colors.Black);
                        }
                    }

                    //=========== DISPLAYING LINE ===================
                    header.Cell().ColumnSpan(11).PaddingVertical(11).BorderBottom(1).BorderColor(Colors.Black);
                });

                //table.Footer(footer =>{var ContractValue = Model.DataDetails.Sum(x => x.ContractValue);
                //    footer.Cell().ColumnSpan(7).Element(CellStyle).AlignLeft().Text("GRAND TOTAL: ").FontSize(14).SemiBold().FontColor(Colors.Green.Darken1);
                //    footer.Cell().Element(CellStyle).AlignRight().Text(ContractValue).FontSize(14).SemiBold().FontColor(Colors.Green.Darken1);});

                //=========== VARIABLE DECLARATION FOR SUB TOTAL GROUP WISE ===================
                string PrevGroupFieldName = "";
                string GroupFieldName = "";

                decimal GroupContractValue = 0;
                decimal GroupPayAmount = 0;
                decimal GroupTDSAmount = 0;
                decimal GroupBalance = 0;
                int GroupItemCount = 0;

                foreach (var item in Model.DataDetails)
                {
                    GroupFieldName = item.CustomerName;

                    if (Model.DataDetails.IndexOf(item) == 0)
                    {
                        //=========== DISPLAYING GROUP FIELD ===================
                        table.Cell().ColumnSpan(11).Element(CellStyle).AlignLeft().Text(GroupFieldName).FontSize(FONT_SIZE_SUB_TOTAL).SemiBold().FontColor(Colors.Blue.Darken2);
                    }
                    else
                    {
                        if (PrevGroupFieldName != GroupFieldName)
                        {
                            if (GroupItemCount > 1)
                            {
                                //=========== DISPLAYING SUB TOTAL ===================
                                table.Cell().ColumnSpan(7).Element(CellStyle).AlignLeft().Text(PrevGroupFieldName + " SUB TOTAL: ").FontSize(FONT_SIZE_SUB_TOTAL).SemiBold().FontColor(Colors.Blue.Darken3);
                                AddField(ref table, GroupContractValue.ToString(), 2);
                                AddField(ref table, GroupPayAmount.ToString(), 2);
                                AddField(ref table, GroupTDSAmount.ToString(), 2);
                                AddField(ref table, GroupBalance.ToString(), 2);
                            }

                            //=========== RESETING SUB TOTAL VARIABLES ===================
                            GroupContractValue = 0;
                            GroupPayAmount = 0;
                            GroupTDSAmount = 0;
                            GroupBalance = 0;
                            GroupItemCount = 0;

                            //=========== DISPLAYING GROUP FIELD ===================
                            table.Cell().ColumnSpan(11).Element(CellStyle).AlignLeft().Text(GroupFieldName).FontSize(FONT_SIZE_SUB_TOTAL).SemiBold().FontColor(Colors.Blue.Darken2);
                        }
                    }
                    //=========== DISPLAYING FIELD ===================
                    table.Cell().Element(CellStyle).Text(Model.DataDetails.IndexOf(item) + 1);
                    AddField(ref table, item.ContractInvoiceNo,1);
                    AddField(ref table, item.ContractInvoiceDate, 1);
                    AddField(ref table, item.ContractOrderNo, 1);
                    AddField(ref table, item.ContractOrderDate, 1);
                    AddField(ref table, item.TransactionNumber, 1);
                    AddField(ref table, item.PaymentReceiptDate, 1);

                    AddField(ref table, item.ContractValue.ToString(), 1, 1);
                    AddField(ref table, item.PayAmount.ToString(), 1, 1);
                    AddField(ref table, item.TDSAmount.ToString(), 1, 1);
                    AddField(ref table, item.Balance.ToString(), 1, 1);

                    //=========== ADDIND VALUES FOR SUB TOTAL ===================
                    GroupContractValue += item.ContractValue;
                    GroupPayAmount += item.PayAmount;
                    GroupTDSAmount += item.TDSAmount;
                    GroupBalance += item.Balance;
                    GroupItemCount++;
                    //=========== ASSIGN LAST GROUP VALUES FOR CHECKING ===================
                    PrevGroupFieldName = GroupFieldName;
                }
            });
        }

        static IContainer CellStyle(IContainer container)
        {
            return container.BorderBottom(1).BorderColor(Colors.Grey.Lighten2).PaddingVertical(5);
        }
        void ComposeComments(IContainer container)
        {
            container.Background(Colors.Grey.Lighten3).Padding(10).Column(column =>
            {
                column.Spacing(5);
                column.Item().Text("Comments").FontSize(14);
                column.Item().Text(Model.Title);
            });
        }
        void ComposeGrandTotal(IContainer container)
        {
            container.Table(table =>
            {                
                table.ColumnsDefinition(columns =>
                {
                    columns.ConstantColumn(25);
                    columns.RelativeColumn(3);
                    columns.RelativeColumn();
                    columns.RelativeColumn();

                    columns.RelativeColumn();
                    columns.RelativeColumn();
                    columns.RelativeColumn();
                    columns.RelativeColumn();

                    columns.RelativeColumn();
                    columns.RelativeColumn();
                    columns.RelativeColumn();
                });

                //=========== GETTING GRAND TOTAL ===================
                var ContractValue = Model.DataDetails.Sum(x => x.ContractValue);
                var PayAmount = Model.DataDetails.Sum(x => x.PayAmount);
                var TDSAmount = Model.DataDetails.Sum(x => x.TDSAmount);
                var Balance = Model.DataDetails.Sum(x => x.Balance);

                //=========== DISPLAYING GRAND TOTAL ===================
                table.Cell().ColumnSpan(7).Element(CellStyle).AlignLeft().Text("GRAND TOTAL: ").FontSize(FONT_SIZE_GRAND_TOTAL).SemiBold().FontColor(Colors.Blue.Darken4);
                AddField(ref table, ContractValue.ToString(), 3);
                AddField(ref table, PayAmount.ToString(), 3);
                AddField(ref table, TDSAmount.ToString(), 3);
                AddField(ref table, Balance.ToString(), 3);
            });
        }

        void AddField(ref TableDescriptor table, string fieldName, int fieldType, int Align = 0)
        {
            if (fieldType == 1)             // NORMAL FIELD VALUE
            {
                if (Align == 1)
                {
                    table.Cell().Element(CellStyle).AlignRight().Text(fieldName).FontSize(FONT_SIZE_FIELD_VALUE).FontColor(Colors.Black);
                }
                else
                {
                    table.Cell().Element(CellStyle).AlignLeft().Text(fieldName).FontSize(FONT_SIZE_FIELD_VALUE).FontColor(Colors.Black);
                }
            }
            else if (fieldType == 2)        // SUB TOTAL FIELD VALUE
            {
                table.Cell().Element(CellStyle).AlignRight().Text(fieldName).FontSize(FONT_SIZE_SUB_TOTAL).SemiBold().FontColor(Colors.Blue.Darken3);
            }
            else if (fieldType == 3)        // GRAND TOTAL FIELD VALUE
            {
                table.Cell().Element(CellStyle).AlignRight().Text(fieldName).FontSize(FONT_SIZE_GRAND_TOTAL).SemiBold().FontColor(Colors.Blue.Darken4);
            }
        }
    }
}

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

The use case you described — repeating group headers and per-page subtotals — is now achievable using [Dynamic Components](https://www.questpdf.com/concepts/code-patterns/dynamic-components.html), which allow you to measure available space, distribute content per page, and calculate subtotals dynamically. A detailed example of per-page subtotals can be found in that documentation.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
